state/unit_test.go

```diff
@@ -71,32 +72,38 @@ func (s *UnitSuite) TestApplication(c *gc.C) {
 
 func (s *UnitSuite) TestUnitStateMutation(c *gc.C) {
 	// Try fetching the state without a state doc present
-	ust, err := s.unit.State()
+	uState, err := s.unit.State()
 	c.Assert(err, gc.IsNil)
-	c.Assert(ust, gc.IsNil, gc.Commentf("expected to receive a nil map when no state doc is present"))
+	c.Assert(uState.State, gc.IsNil, gc.Commentf("expected to receive a nil map when no state doc is present"))
+	c.Assert(uState.UniterState, gc.Equals, "")
 
 	// Set initial state; this should create a new unitstate doc
 	initialState := map[string]string{
 		"foo":          "bar",
 		"key.with.dot": "must work",
 		"key.with.$":   "must work to",
 	}
-	err = s.unit.SetState(initialState)
+	initialUniterState := "testing"
+	err = s.unit.SetState(unit.UnitState{State: initialState, UniterState: initialUniterState})
 	c.Assert(err, gc.IsNil)
 
 	// Read back initial state
-	ust, err = s.unit.State()
+	uState, err = s.unit.State()
 	c.Assert(err, gc.IsNil)
-	c.Assert(ust, gc.DeepEquals, initialState)
+	c.Assert(uState.State, gc.DeepEquals, initialState)
+	c.Assert(uState.UniterState, gc.Equals, initialUniterState)
 
 	// Mutate state again with an existing state doc
 	newState := map[string]string{"foo": "42"}
-	err = s.unit.SetState(newState)
+	err = s.unit.SetState(unit.UnitState{State: newState})
 	c.Assert(err, gc.IsNil)
 
-	ust, err = s.unit.State()
+	uState, err = s.unit.State()
 	c.Assert(err, gc.IsNil)
-	c.Assert(ust, gc.DeepEquals, newState)
+	c.Assert(uState.State, gc.DeepEquals, newState)
+	// UniterState should never be removed.  Since SetState had an empty string,
+	// ensure the initialUniterState is still there.
+	c.Assert(uState.UniterState, gc.Equals, initialUniterState)
 }
 
 func (s *UnitSuite) TestUnitStateNopMutation(c *gc.C) {
```

Comment:
Can you expand this test (or duplicate if you prefer) and make sure that calling SetState with the same data (uniter state, relations, storage) results in ErrNoOperations.

Reply:
done